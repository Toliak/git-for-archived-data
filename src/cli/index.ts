import {
    createWatcher,
    formatRawData,
    packArchive,
    unpackArchive,
} from '../core/index.js';
import {
    configToString,
    mapConfigToPromises,
    readConfig,
} from '../config/index.js';

import { ArgumentParser } from 'argparse';
import { GitForArchivedData } from '../config/types.js';
import fs from 'fs';
import path from 'path';
import { askToUserConfirm } from './utils.js';
import { initializeDirectory } from './initializer.js';

const parser = new ArgumentParser({
    description: 'Git for Archived Data',
});

const enum PackerActionType {
    Pack = 'pack',
    Unpack = 'unpack',
    Watch = 'watch',
    Initialize = 'init',
}

const packerActions: PackerActionType[] = [
    PackerActionType.Pack,
    PackerActionType.Unpack,
    PackerActionType.Watch,
    PackerActionType.Initialize,
];

parser.add_argument('-a', '--action', {
    dest: 'action',
    choices: packerActions as string[],
    help: 'Archive action',
    required: true,
});

parser.add_argument('-c', '--config', {
    dest: 'config',
    default: 'git-for-archived-data.json',
    help: 'Configuration path',
    required: false,
});

interface CliArguments {
    action: PackerActionType;
    config: string;
}

type ActionFunction = (this: void, config: GitForArchivedData) => void;
const actionFunctions: Record<PackerActionType, ActionFunction> = {
    [PackerActionType.Unpack]: async config => {
        const promises = mapConfigToPromises(config, async item => {
            const archivePath = path.resolve(item.archive.path);
            const rawPath = path.resolve(item.raw.path);

            await unpackArchive(archivePath, rawPath);

            if (item.raw.applyPrettier) {
                await formatRawData(rawPath, '.');
            } else {
                console.log('Skipped prettier');
            }
        });
        return Promise.all(promises);
    },

    [PackerActionType.Pack]: async config => {
        const userConfirmation = await askToUserConfirm();
        if (!userConfirmation) {
            console.error('\x1b[31m⛊\x1b[0m Aborted');
            return;
        }

        const promises = mapConfigToPromises(config, async item => {
            await packArchive(
                item.raw.path,
                item.archive.path,
                item.archive.format as never,
            );
        });
        return Promise.all(promises);
    },

    [PackerActionType.Watch]: async config => {
        const rawPathsToWatch = config.items
            .filter(it => it.archive.watching)
            .map(it => path.resolve(it.archive.path));

        if (rawPathsToWatch.length === 0) {
            console.error('\x1b[31m⛊\x1b[0m Nothing to watch');
            return;
        }

        console.log(
            `Paths to watch: \n${rawPathsToWatch
                .map(it => `- \x1b[34m${it}\x1b[0m`)
                .join('\n')}`,
        );

        createWatcher(config);
    },

    [PackerActionType.Initialize]: async () => {
        await initializeDirectory('.');
        console.log(`\x1b[32m♦\x1b[0m ` + 'Initialized');
    },
};

// TODO: resolve paths preliminary

export async function parseArguments(args: string[]): Promise<void> {
    const parseArgs: CliArguments = parser.parse_args(args);

    const configPath = path.resolve(parseArgs.config);
    if (!fs.existsSync(configPath)) {
        console.error(`\x1b[31m⛊\x1b[0m The config "${configPath}" not found`);
        return;
    }

    if (parseArgs.action === PackerActionType.Initialize) {
        // Do not read config for initialization
        await actionFunctions[parseArgs.action](null as never);
        return;
    }

    const config = readConfig(configPath);
    console.log(configToString(config));

    const actionFunction = actionFunctions[parseArgs.action];
    if (!actionFunction) {
        console.error(`\x1b[31m⛊\x1b[0m Unknown action: "${parseArgs.action}"`);
        return;
    }

    await actionFunction(config);
    return;
}
