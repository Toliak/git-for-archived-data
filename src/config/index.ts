import * as JSON5 from 'json5';
import { Validator, ValidatorResult } from 'jsonschema';
import { GitForArchivedData } from './types';
import path from 'path';
import * as fs from 'fs';

class ConfigReaderError extends Error {}

export function validateConfig(data: unknown): ValidatorResult {
    const schemaPath = path.resolve(
        __dirname,
        '../../git-for-archived-data.schema.json',
    );
    const schemaContent = fs.readFileSync(schemaPath, 'utf8');
    const schema = JSON5.parse(schemaContent);

    const validator = new Validator();
    return validator.validate(data, schema, {
        throwError: false,
    });
}

export function readConfig(filepath: string): GitForArchivedData {
    const content = fs.readFileSync(filepath, 'utf8');
    const metaArray = JSON5.parse(content);

    for (let i = 0; i < metaArray.length; ++i) {
        const meta = metaArray[i];

        const validationResult = validateConfig(meta);
        if (validationResult.errors.length > 0) {
            console.log(
                `Error happen while validating "${filepath}". In part: ${i}`,
            );
            throw new ConfigReaderError(validationResult.toString());
        }
    }

    return metaArray;
}
