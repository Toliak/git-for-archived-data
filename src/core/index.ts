export { unpackArchive, packArchive } from './packer';
export { formatRawData } from './prettier';
