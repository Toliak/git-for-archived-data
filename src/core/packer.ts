import fs from 'fs';
import { PullStream } from 'unzipper';
import * as unzip from 'unzipper';
import archiver, { Format } from 'archiver';

export async function unpackArchive(
    archivePath: string,
    rawPath: string,
): Promise<void> {
    const pipe = fs.createReadStream(archivePath);

    return new Promise<void>(function (resolve, reject) {
        const unzipPipe: PullStream = pipe.pipe(
            unzip.Extract({ path: rawPath }),
        );
        unzipPipe.on('error', err => {
            console.error(
                '\x1b[31m⛊\x1b[0m ' +
                '\x1b[31mError happen while unpacking\x1b[0m',
            );
            reject(err);
        });
        unzipPipe.on('close', () => {
            console.log('\x1b[32m♦\x1b[0m Unpacking complete\x1b[0m');
            resolve();
        });
    });
}

export async function packArchive(
    rawPath: string,
    archivePath: string,
    format: Format,
): Promise<void> {
    const archive = archiver(format, { zlib: { level: 9 } });
    const stream = fs.createWriteStream(archivePath);

    return new Promise((resolve, reject) => {
        archive
            .directory(rawPath, false)
            .on('error', err => reject(err))
            .pipe(stream);

        stream.on('close', () => resolve());
        archive.finalize();
    });
}
