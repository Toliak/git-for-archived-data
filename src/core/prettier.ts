import * as prettier from 'prettier';
import fs from 'fs';
import glob from 'glob';
import path from 'path';

class PrettierError extends Error {}

async function formatRawDataFile(
    fullPath: string,
    options: prettier.Options,
): Promise<void> {
    const info = await prettier.getFileInfo(fullPath, {
        resolveConfig: true,
    });
    if (info.ignored) {
        console.log(`Skipping ignored ${fullPath}`);
        return;
    }
    if (!info.inferredParser) {
        console.log(`No parser found for ${fullPath}`);
        return;
    }

    const source = fs.readFileSync(fullPath, { encoding: 'utf8' });
    const result = await prettier.format(source, {
        ...options,
        filepath: fullPath,
    });
    fs.writeFileSync(fullPath, result, { encoding: 'utf8' });
}

export async function formatRawData(
    directoryPath: string,
    prettierConfigPath: string,
): Promise<void> {
    const options = await prettier.resolveConfig(prettierConfigPath);
    if (!options) {
        console.error(
            '\x1b[31m⛊\x1b[0m ' +
                ' Cannot find prettier config file' +
                '(Check .prettierrc.json file in the root of the project)',
        );
        throw new PrettierError();
    }

    const fullPath = path.resolve(directoryPath);
    if (!fs.lstatSync(fullPath).isDirectory()) {
        console.log(
            `\x1b[32m♦\x1b[0m ` +
                `Prettier applied to the file \x1b[34m${directoryPath}\x1b[0m`,
        );
        return formatRawDataFile(fullPath, options);
    }

    const promises: Promise<void>[] = [];

    for (const filePath of glob.sync('**', {
        cwd: directoryPath,
    })) {
        const fullFilePath = path.join(directoryPath, filePath);
        if (fs.lstatSync(fullFilePath).isDirectory()) {
            continue;
        }

        promises.push(formatRawDataFile(fullFilePath, options));
    }

    await Promise.all(promises);
    console.log(
        `\x1b[32m♦\x1b[0m ` +
            `Prettier applied to the files in \x1b[34m${directoryPath}\x1b[0m`,
    );
}
