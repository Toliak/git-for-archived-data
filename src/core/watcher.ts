import chokidar from 'chokidar';
import { GitForArchivedData, Item as ConfigItem } from '../config/types';
import { unpackArchive } from './packer';
import { formatRawData } from './prettier';

interface WatcherContext {
    lastBuildTime: number;
}

const WATCHER_COOLDOWN = 1;

async function listenerTemplate(
    path: string,
    context: WatcherContext,
    mapping: Readonly<Record<string, ConfigItem>>,
): Promise<void> {
    // If current datetime - lastBuild datetime < DELTA => do not build again!
    console.log(
        `> Updated \x1b[34m${path}\x1b[0m ` +
            `\x1b[35m${new Date().toISOString()}\x1b[0m`,
    );
    const currentTimestamp = new Date().getTime();
    if (currentTimestamp - context.lastBuildTime < WATCHER_COOLDOWN) {
        console.log(`> Build \x1b[31mskipped due to cooldown\x1b[0m`);
        return;
    }

    const config = mapping[path];
    if (!config) {
        console.log(
            `> Error \x1b[34m${path}\x1b[0m ` +
                `\x1b[35m${new Date().toISOString()}\x1b[0m`,
        );
        return;
    }
    await unpackArchive(config.raw.path, config.archive.path);
    if (config.raw.applyPrettier) {
        await formatRawData(config.raw.path, '.prettierrc.json');
    }

    context.lastBuildTime = new Date().getTime();
}

export default function (
    config: Readonly<GitForArchivedData>,
): chokidar.FSWatcher {
    console.log('> Watcher \x1b[32mstarted\x1b[0m');

    const context: WatcherContext = {
        lastBuildTime: 0,
    };

    const archivesMap = Object.fromEntries(
        config.items
            .filter(v => v.archive.watching)
            .map(v => [v.archive.path, v]),
    );

    const eventClosure: (path: string) => void = p =>
        listenerTemplate(p, context, archivesMap);
    return chokidar
        .watch(Object.keys(archivesMap), {
            ignoreInitial: true,
        })
        .on('add', eventClosure)
        .on('change', eventClosure);
}
