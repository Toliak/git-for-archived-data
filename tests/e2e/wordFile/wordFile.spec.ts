import * as path from 'path';
import {GitForArchivedData} from '../../../src/config/types';
import {readConfig} from '../../../src/config';
import {formatRawData, packArchive, unpackArchive} from '../../../src/core/';
import {
    getDefaultConfigPath,
    removeDirectoryIfExists,
    targetDirectoryFilesLengthTest,
    targetFileCheckTest,
} from '../mixins';

interface TestEnvContext {
    config: GitForArchivedData | null;
}

/**
 * Do not execute sub-tests separately
 */
describe('Case MS Office Word file unpack and pack', () => {
    const targetDirectory = 'tests/e2e/wordFile';

    const configPath = getDefaultConfigPath(targetDirectory);
    const prettierConfigPath = '.prettierrc.json';

    const rawDirectoryPath = path.join(__dirname, 'SampleDocumentResult');
    const resultArchivePath = path.join(__dirname, 'SampleDocumentResult.docx');

    const context: TestEnvContext = {
        config: null,
    };

    removeDirectoryIfExists(rawDirectoryPath);

    it('ReadConfig', async () => {
        context.config = readConfig(configPath);
        expect(context.config.items).toHaveLength(1);
    });

    it('Unpack word file', async () => {
        expect(context.config).not.toBeNull();
        // eslint-disable-file @typescript-eslint/no-non-null-assertion
        const item = context.config!.items[0]!;

        await unpackArchive(
            path.join(__dirname, item.archive.path),
            path.join(__dirname, item.raw.path),
        );
    });

    targetFileCheckTest(rawDirectoryPath, true);
    targetDirectoryFilesLengthTest(rawDirectoryPath, 4);

    targetFileCheckTest(path.join(rawDirectoryPath, 'word'), true);
    targetFileCheckTest(path.join(rawDirectoryPath, '_rels'), true);
    targetFileCheckTest(
        path.join(rawDirectoryPath, '[Content_Types].xml'),
        true,
    );
    targetFileCheckTest(path.join(rawDirectoryPath, 'docProps'), true);

    it('Format raw data', async () => {
        await formatRawData(path.join(rawDirectoryPath, 'word'), prettierConfigPath);
        await formatRawData(path.join(rawDirectoryPath, '_rels'), prettierConfigPath);
        await formatRawData(path.join(rawDirectoryPath, 'docProps/app.xml'), prettierConfigPath);
        await formatRawData(path.join(rawDirectoryPath, 'docProps/core.xml'), prettierConfigPath);
        await formatRawData(path.join(rawDirectoryPath, '[Content_Types].xml'), prettierConfigPath);
    });

    it('Pack word file', async () => {
        expect(context.config).not.toBeNull();
        const item = context.config!.items[0]!;

        await packArchive(
            path.join(__dirname, item.raw.path),
            resultArchivePath,
            'zip',
        );
    });

    targetFileCheckTest(resultArchivePath, true);
});
