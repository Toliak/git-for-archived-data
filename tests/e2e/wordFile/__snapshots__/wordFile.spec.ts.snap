// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Case MS Office Word file unpack and pack The target file "tests/e2e/wordFile/SampleDocumentResult/[Content_Types].xml" exists 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Types
    xmlns="http://schemas.openxmlformats.org/package/2006/content-types"
><Default
        Extension="rels"
        ContentType="application/vnd.openxmlformats-package.relationships+xml"></Default><Default
        Extension="xml"
        ContentType="application/xml"></Default><Override
        PartName="/word/document.xml"
        ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml"></Override><Override
        PartName="/word/styles.xml"
        ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.styles+xml"></Override><Override
        PartName="/word/settings.xml"
        ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.settings+xml"></Override><Override
        PartName="/word/webSettings.xml"
        ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.webSettings+xml"></Override><Override
        PartName="/word/fontTable.xml"
        ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.fontTable+xml"></Override><Override
        PartName="/word/theme/theme1.xml"
        ContentType="application/vnd.openxmlformats-officedocument.theme+xml"></Override><Override
        PartName="/docProps/core.xml"
        ContentType="application/vnd.openxmlformats-package.core-properties+xml"></Override><Override
        PartName="/docProps/app.xml"
        ContentType="application/vnd.openxmlformats-officedocument.extended-properties+xml"></Override></Types>
"
`;

exports[`Case MS Office Word file unpack and pack The target file "tests/e2e/wordFile/SampleDocumentResult/word/_rels/document.xml.rels" exists 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Relationships
    xmlns="http://schemas.openxmlformats.org/package/2006/relationships"
><Relationship
        Id="rId3"
        Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/webSettings"
        Target="webSettings.xml"></Relationship><Relationship
        Id="rId2"
        Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/settings"
        Target="settings.xml"></Relationship><Relationship
        Id="rId1"
        Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles"
        Target="styles.xml"></Relationship><Relationship
        Id="rId5"
        Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/theme"
        Target="theme/theme1.xml"></Relationship><Relationship
        Id="rId4"
        Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/fontTable"
        Target="fontTable.xml"></Relationship></Relationships>
"
`;

exports[`Case MS Office Word file unpack and pack The target file "tests/e2e/wordFile/SampleDocumentResult/word/document.xml" exists 1`] = `
"<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<w:document
    xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas"
    xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex"
    xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:o="urn:schemas-microsoft-com:office:office"
    xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"
    xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math"
    xmlns:v="urn:schemas-microsoft-com:vml"
    xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing"
    xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing"
    xmlns:w10="urn:schemas-microsoft-com:office:word"
    xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"
    xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml"
    xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml"
    xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex"
    xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup"
    xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk"
    xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml"
    xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape"
    mc:Ignorable="w14 w15 w16se wp14"
><w:body><w:p w:rsidR="00E86CBA" w:rsidRDefault="006D0117"><w:r><w:t
                >Test data</w:t></w:r><w:bookmarkStart
                w:id="0"
                w:name="_GoBack"></w:bookmarkStart><w:bookmarkEnd
                w:id="0"></w:bookmarkEnd></w:p><w:sectPr
            w:rsidR="00E86CBA"
        ><w:pgSz w:w="12240" w:h="15840"></w:pgSz><w:pgMar
                w:top="1134"
                w:right="850"
                w:bottom="1134"
                w:left="1701"
                w:header="720"
                w:footer="720"
                w:gutter="0"></w:pgMar><w:cols w:space="720"></w:cols><w:docGrid
                w:linePitch="360"></w:docGrid></w:sectPr></w:body></w:document>
"
`;
